### ISSUE TITLE
Add candidate sampling methods

### ISSUE THREAD

**Issue by danielwatson6** [danielwatson6/skip-thoughts#9]
The biggest overhead in the model is the softmax operation that converts the decoders' output shapes to the vocabulary size, which is huge (even 20,000 words results in too many out-of-vocabulary words). Different [candidate sampling methods](https://www.tensorflow.org/extras/candidate_sampling.pdf) should be supported to tackle this.
